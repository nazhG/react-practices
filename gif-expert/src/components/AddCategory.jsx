import { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ onNewCategory }) => {

    const [input, setInput] = useState('');

    const handleInput = (e) => {
        setInput(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        console.log(input);
        
        if (input.trim().length < 2) {
            return;
        }

        onNewCategory(input);
        setInput('');
    }

    return (
        <form onSubmit={handleSubmit} name="form">
            <input 
                type="text" 
                name="category" 
                id="cat" 
                value={input} 
                onChange={handleInput}
            />
        </form>
    )
}

AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired
}