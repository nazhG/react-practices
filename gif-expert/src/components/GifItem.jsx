import React from 'react'
import PropTypes from 'prop-types';

export const GifItem = ({id, url, title }) => {
  return (
    <img 
      key={id} 
      src={url} 
      alt={title} 
      className="gif-item"
    />
  )
}

//npm i prop-types
GifItem.propTypes = {
  id: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}