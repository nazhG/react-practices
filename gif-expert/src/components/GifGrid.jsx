import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";

export const GifGrid = ({ category }) => {
    
    const { images, loading, error } = useFetchGifs(category);

    return (
        <>
            <h1>{category}</h1>
            <div className="gif-grid">
                {/* {images.map(({id, images:{downsized_medium:{url}}, title}) => ( */}
                {images.map(image => (
                    /* !loading && (<h2>Cargando ...</h2>) */
                    loading?
                    (<h2>Cargando ...</h2>):
                    GifItem({ ...image })
                ))}
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
