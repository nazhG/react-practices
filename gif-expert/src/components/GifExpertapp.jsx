import { useState } from "react";
import { AddCategory } from "./AddCategory";
import { GifGrid } from "./GifGrid";

export const GifExpertapp = () => {
  const [categories, setCategories] = useState(["One Punch", "Dragon Ball"]);

  const onAddCategory = (cat) => {
    if (categories.indexOf(cat) === -1) {
      setCategories([cat, ...categories]);
    }
  };

  return (
    <>
      <h1>GifExpertapp</h1>

      <AddCategory onNewCategory={(cat) => onAddCategory(cat)} />

      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};
