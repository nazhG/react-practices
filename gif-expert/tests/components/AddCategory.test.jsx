import { fireEvent, render, screen } from "@testing-library/react";
import { AddCategory } from "../../src/components/AddCategory";

describe('Puebas en addCategory', () => { 
    
    // screen.debug();

   
    test('debe de cambiar la caja de texto', () => {
        render(<AddCategory onNewCategory={ () => {} } />);
        const input = screen.getByRole('textbox');
        fireEvent.input(input, { target: { value: 'Hola mundo' } });
        expect(input.value).toBe('Hola mundo');
    })

    test('debe de llamar el submit', () => {
        const onNewCategory = jest.fn();

        const inputValue = 'Hola mundo';
        render(<AddCategory onNewCategory={ onNewCategory } />);

        const input = screen.getByRole('textbox');
        const from = screen.getByRole('form');

        fireEvent.input(input, { target: { value: inputValue } });
        fireEvent.submit(from);

        expect(input.value).toBe('');

        //toHaveBeenCalled()
        expect(onNewCategory).toHaveBeenCalledTimes(1);
        expect(onNewCategory).toHaveBeenCalledWith(inputValue);
    })

    test('no debe de llamar el submit', () => {
        const onNewCategory = jest.fn();

        const inputValue = '';
        render(<AddCategory onNewCategory={ onNewCategory } />);

        const input = screen.getByRole('textbox');
        const from = screen.getByRole('form');

        fireEvent.input(input, { target: { value: inputValue } });
        fireEvent.submit(from);
    

        expect(input.value).toBe('');

        //toHaveBeenCalled()
        expect(onNewCategory).not.toHaveBeenCalled();
    })

});