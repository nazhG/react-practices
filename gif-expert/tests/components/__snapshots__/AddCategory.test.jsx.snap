// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puebas en addCategory debe de mostrar el componente correctamente 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <input
          id="cat"
          name="category"
          type="text"
          value=""
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <input
        id="cat"
        name="category"
        type="text"
        value=""
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
